Add CMakeLists alongside existing makefiles

reuk opened the pull request:
By adding a CMake build alongside the makefile build, we can continue supporting existing users of CBMC, while also supporting development in a wider range of IDEs, with faster build back-ends and better integration with clang-based static-analysis and code completion tools (clang-tidy and YCM).

As long as both builds are tested in CI, we don't need to worry about the build systems getting out-of-sync. I don't think having two build systems will add maintenance overhead: I've been maintaining this build for the last three months or so, and the only time I've had to update it was when unit tests were added. This is possible because I've set the build to automatically pick-up all the source files in each directory containing a CMakeLists.txt (in fact, some of the source files *don't build* and should probably be removed).

reuk commented:
> Is there a way to verify that the make files generated by this are equivalent (or perhaps generate them from this, and replace the current Makefiles with the generated ones)

The generated makefiles will *not* be equivalent. CMake generates machine-dependent makefiles (the point being that it abstracts over different machine configurations itself, and generates tailor-made makefiles). Therefore, it's not viable to replace the makefile build with cmake-generated makefiles.

> if there are any gotchas or weird non cmake-standard things in here, could they be documented (perhaps there are none).

There's nothing too weird in here, but if this is adopted I'll add a "getting started with CMake" to the wiki.

tautschnig commented:
Please add documentation to `COMPILING` that explains how to use the build system, if there are any changes over the current state.

tautschnig commented:
The CI jobs should do some form of checking beyond "it compiles." This might either be enabling the regression and/or unit tests, or a simple comparison of the binaries (which ought to be identical, after all).

reuk commented:
I just thought of something that won't work yet: the CBMC executable will be built out-of-source, so the regression tests won't be able to find it. Given that the user might put the build anywhere, it's not practical to hard-code the path. Maybe for now I should just explain that users need to symlink the CBMC executable into `src/cbmc`, but there's probably a better solution I haven't thought of.

tautschnig commented on COMPILING:
```diff
@@ -183,6 +183,40 @@ can be used for building with MSBuild.  Note that you still need to run
+---------------------------------
+
+There is an experimental build based on CMake instead of hand-written
+makefiles. You may prefer to use this, as it should work on a wider variety of
```
Omit "You may prefer to use this" -- this isn't marketing.

tautschnig commented on COMPILING:
```diff
@@ -183,6 +183,40 @@ can be used for building with MSBuild.  Note that you still need to run
+   have a package for CMake, and Mac users can get it through homebrew.
+   Windows users should download it manually from cmake.org.
+
+1) Navigate to the CBMC directory in a terminal.
```
Get the instructions in line with the above instructions: precise commands, no fuffing about a "CBMC directory" (no idea what that is).

tautschnig commented on COMPILING:
```diff
@@ -183,6 +183,40 @@ can be used for building with MSBuild.  Note that you still need to run
+
+1) Navigate to the CBMC directory in a terminal.
+
+2) Create somewhere to store your build:
```
Be precise in your wording: "Create a directory", not "somewhere"

tautschnig commented on COMPILING:
```diff
@@ -183,6 +183,40 @@ can be used for building with MSBuild.  Note that you still need to run
+   build back-ends. You can use clang (for example) by adding the argument
+   `-DCMAKE_CXX_COMPILER=clang++` to the command line. You can also tell
+   CMake to generate IDE projects by supplying the `-G` flag.
+   Run `cmake -G` for a comprehensive list of supported back-ends.
```
Thank you for the detailed instructions. Yet: how to set compiler options? For example, how to change optimisation, how to set additional linking flags? The bits currently in config.inc must be clearly documented here.

tautschnig commented on src/ansi-c/CMakeLists.txt:
```diff
@@ -0,0 +1,70 @@
+    ${CMAKE_CURRENT_BINARY_DIR}/gcc_builtin_headers_ubsan.inc
+)
+
+generic_includes(ansi-c)
```
Where is `make library_check` implemented?

reuk replied:
It's not yet. I'll look at that now.

tautschnig commented on src/CMakeLists.txt:
```diff
@@ -0,0 +1,147 @@
+)
+add_dependencies(minisat2-condensed minisat2-extern)
+
+#   Override add_executable to automatically sign the target on OSX.
```
Where are universal binaries and the MacOS version implemented?

reuk replied:
That's handled automatically. By default, the build will be for the architecture of the machine running the build. However, you can also supply the flag `-DCMAKE_OSX_ARCHITECTURES=i386;x86_64` to build fat binaries.

tautschnig replied:
Will `-mmacosx-version-min=10.9` also be supported? Note that this ought to be the default as we have had problems in the past with users being unable to run the tool on those older MacOS versions.

Essentially, whatever is currently configured in `common` and `config.inc` ought to be the default for the CMake builds as well.

tautschnig commented:
> I just thought of something that won't work yet: the CBMC executable will be built out-of-source, so the regression tests won't be able to find it. Given that the user might put the build anywhere, it's not practical to hard-code the path. Maybe for now I should just explain that users need to symlink the CBMC executable into src/cbmc, but there's probably a better solution I haven't thought of.

Shouldn't the regression tests (and unit tests!) also be run from within the "build" (or whatever the user chosen name) folder? That is, all the artifacts generated by running regression tests ought to end up within the "build" folder, not underneath the "regression" folder as is currently the case.

reuk commented:
> Shouldn't the regression tests (and unit tests!) also be run from within the "build" (or whatever the user chosen name) folder? That is, all the artifacts generated by running regression tests ought to end up within the "build" folder, not underneath the "regression" folder as is currently the case.

They *should*, and this is the case for the unit tests. I'll try to fix this for regression tests too.

reuk commented:
I've been hacking around with the regression tests. The ones with custom driver scripts are a bit awkward, because the scripts have to be adjusted to drive arbitrary executables instead of hard-coded ones. Other than that, it seems to work (on my machine). If it gets through CI I'll squash the commits appropriately.

thk123 commented:
@reuk  Any idea why the windows build is failing? 

```
preprocessing has failed
```

is an exception message caused by running what looks like the native compiler and it returning a non-zero code. Not immediately obvious why this would be caused by this, but perhaps some change to the regressions folder has caused this?

reuk commented:
OK, the build is passing and running units + regressions! AFAIK the only things left are:
- Add the `library_check` target
- Ensure `goto-cc` is named `goto-cl` on Windows
- Tidy and squash the history

reuk commented:
Squashed and ready for re-review @tautschnig (and anyone else interested in the progression of this PR).

Some notes:
- library_check doesn't pass currently, even in the makefile build. I've added a target replicating the behavior in the CMake build, but of course this fails too.
- the appveyor build was doing some nasty stuff where it edited test driver scripts using `sed`, to inject custom executable names/behaviors. I think it's nicer to do this with a conditional inside the scripts themselves, so I've changed that too

tautschnig commented:
#1243 is in need of review and merge - this would make library_check actually pass.

tautschnig commented on src/ansi-c/check_library.sh:
```diff
@@ -0,0 +1,14 @@
+    sed -i 's/__sync_/s&/' __libcheck.c
+    sed -i 's/__noop/s&/' __libcheck.c
+    cc -std=gnu99 -E -include library/cprover.h -D__CPROVER_bool=_Bool -D__CPROVER_thread_local=__thread -DLIBRARY_CHECK -o __libcheck.i __libcheck.c
+    cc -S -Wall -Werror -pedantic -Wextra -std=gnu99 __libcheck.i -o __libcheck.s -Wno-unused-label -Wno-uninitialized
```
Would you mind incorporating the changes that #1243 made to ansi-c/Makefile right away? Otherwise we end up in inconsistency-hell.

reuk replied:
How soon will #1243 be merged? If the answer is "soon", I'm inclined to wait until it's in and then to rebase this PR. I could then factor the check-library stuff into a script which is called from both the makefiles and the CMake build.

tautschnig replied:
#1243 is now merged :-)

tautschnig commented on src/goto-analyzer/CMakeLists.txt:
```diff
@@ -0,0 +1,35 @@
+    ${CMAKE_CURRENT_SOURCE_DIR}/goto_analyzer_main.cpp
+
+    # This doesn't build
+    ${CMAKE_CURRENT_SOURCE_DIR}/xml_binaries/xml_goto_function.cpp
```
Please file a PR removing this file.

reuk replied:
To clarify, are you suggesting removing *all* files which aren't mentioned in makefiles and don't build, or just the two that you've pointed out?

tautschnig replied:
I'm all up for removing all files that are never mentioned in Makefiles!

reuk replied:
OK, the pr is #1321

tautschnig commented on src/goto-instrument/CMakeLists.txt:
```diff
@@ -0,0 +1,36 @@
+    ${CMAKE_CURRENT_SOURCE_DIR}/goto_instrument_main.cpp
+
+    # This doesn't build
+    ${CMAKE_CURRENT_SOURCE_DIR}/accelerate/linearize.cpp
```
Please file a PR removing this file.

reuk commented:
Added a PR fixing the library_check target now that #1243 is merged. Left in a separate commit for easy review for now. I'll squash it once the change has been reviewed.

reuk commented:
Note to self: I need to set up a persistent travis cache for the CMake builds

reuk commented:
This is failing because the 32-bit OS X build is broken: #1330

tautschnig commented:
PR subject changed for visibility.

reuk commented:
The 32-bit compatibility problem is blocked at the moment (#1353). Therefore, I will remove the 32-bit mac build from CI for the time being. This brings CMake in line with the functionality present in the makefile build (which is also broken on 32-bit platforms). Once the 32-bit build is fixed, it can be enabled very simply by changing `'-DCMAKE_OSX_ARCHITECTURES=x86_64'` to `'-DCMAKE_OSX_ARCHITECTURES=i386;x86_64'` in the Travis file.

reuk commented:
I've implemented all suggested changes, other than those blocked by other issues/PRs (#1353 and #1321). If I clean the CMake files up once those issues are resolved, is this currently in a merge-able state?

tautschnig commented:
> If I clean the CMake files up once those issues are resolved, is this currently in a merge-able state?

From my point of view: yes. Yet I'd hope for @peterschrammel and @kroening to weigh in as well.

peterschrammel commented on src/solvers/CMakeLists.txt:
```diff
@@ -0,0 +1,85 @@
+#   TODO Specify which solver to use (minisat2 etc.)
```
Could you try using glucose and document the changes required to get it to run?

reuk replied:
I'll do that now.

peterschrammel commented on .travis.yml:
```diff
@@ -151,6 +151,34 @@ jobs:
+          packages:
+            - g++-5
+      install:
+        - cmake -H. -Bbuild '-DCMAKE_BUILD_TYPE=Release' '-DCMAKE_CXX_FLAGS_RELEASE=-O3' '-DCMAKE_CXX_FLAGS=-Wall -Wpedantic -Werror' '-DCMAKE_CXX_COMPILER=g++-5'
```
Do we compile with -O3 now?

reuk replied:
The default "Release" flags for clang++ and g++ are `-O3 -DNDEBUG`. We have to keep assertions enabled, which is why I've manually overridden the flags. I can switch to `-O2` too if you'd like.

tautschnig replied:
I'd say for a start the Release build ought to be using an as similar set of flags as the current build as possible. Is it possible to make this the default in the CMake configuration rather than requiring overrides on the command line?

reuk commented:
I've addressed the last few comments.

The glucose build can now be enabled by supplying the flag `'-Dsat_impl=glucose'` to the initial call to CMake (there's an example in the travis config). The default sat impl remains minisat2. As a byproduct of this, I updated satcheck.h and config.inc a bit, so that glucose can be enabled from the makefiles just by modifying config.inc with the path to glucose. There's no longer a need to update the defines in satcheck.h.

Also, build flags should match the makefile flags more closely by default. This also allows the travis config to be simplified.

reuk commented:
I've also simplified the CMake files based on #1321.

reuk commented:
Some of the regression tests fail when glucose is used. Switching this off for now.

reuk commented:
Mac builds on Travis are broken right now, but I just tried this on a real-life macbook (build with the settings from the Travis config and then run tests) and I can confirm it works. It'd be nice to get this merged before the end of the week, any chance of that happening?

tautschnig commented:
Some of the regression tests fail when glucose is used. Switching this off for now.

Can this be captured in an issue, possibly also with a note on what the mismatch is?

Thanks!

reuk commented:
I've added an issue, #1366, describing the problem

reuk commented:
@peterschrammel are you happy with this now?

peterschrammel commented:
@reuk, is this mergeable or do you plan further additions for now?

reuk commented:
Yes, it's mergeable
